### PR TITLE
Switch to Scroll-owned Goldilocks repository

### PR DESCRIPTION
**Pull request by matthiasgoergens**
Also remove manual `Ord` implementation for `Expression` and use a derived implementation that does the same as the manual one used to do.

### CODE REVIEW

**Comment by yczhangsjtu** [comment]
For the `Ord`, I don't know why is there a custom implementation in the first place? Was there any special considerations?

**Comment by matthiasgoergens** [comment]
> For the `Ord`, I don't know why is there a custom implementation in the first place? Was there any special considerations?

I only added derive Ord to the upstream repo (which is now scroll owned) recently.

**Comment by matthiasgoergens** [comment]
@yczhangsjtu Please tell me, if there's anything left you want me to do before you can approve?  (Or let me know, if you just feel uncomfortable with the responsibility.)

**Comment by yczhangsjtu** [comment]
> @yczhangsjtu Please tell me, if there's anything left you want me to do before you can approve? (Or let me know, if you just feel uncomfortable with the responsibility.)

I do feel uneasy since the default implementation may introduce a different order on the types from here, and I don't know if that might break anything. But I figure it should be okay if the tests pass.

**Comment by yczhangsjtu** [comment]
> > For the `Ord`, I don't know why is there a custom implementation in the first place? Was there any special considerations?
> 
> I only added 'derive Ord' to the upstream repo (which is now scroll owned) recently.
> 
> In general, we should do more upstream fixes, instead of just working around them.

So the reason for this custom implementation was that `Goldilocks` didn't implement `Ord`, right?

**Review by yczhangsjtu — APPROVED** [review]

**Comment by matthiasgoergens** [comment]
> So the reason for this custom implementation was that `Goldilocks` didn't implement `Ord`, right?

Well, that's my best guess from reading the code and digging through the git history.  There was lots of other weirdness that I can't explain, though.  For example, why all the weird cases analysis, instead of [`then`](https://doc.rust-lang.org/std/cmp/enum.Ordering.html#method.then) or [`then_with`](https://doc.rust-lang.org/std/cmp/enum.Ordering.html#method.then_with).

**Comment by matthiasgoergens** [comment]
> I do feel uneasy since the default implementation may introduce a different order on the types from here, and I don't know if that might break something. But I figure it should be okay if the tests pass.

That's a fair point to make.  Do you want me to investigate, or are you happy with the existing tests?

File: ceno_zkvm/src/expression.rs
Comment (naure) [inline ceno_zkvm/src/expression.rs:25]: Nice!

**Review by naure — APPROVED** [review]
Are there any other differences besides Ord?

**Comment by matthiasgoergens** [comment]
> Are there any other differences besides Ord?

Not really.

**Comment by naure** [comment]
That weird custom Ord is used to deduplicate commutative operations (b*a == a*b).

It can be replaced with the default order now.

**Comment by matthiasgoergens** [comment]
> That weird custom Ord is used to deduplicate commutative operations (`b*a == a*b`).
> 
> It can be replaced with the default order now.

I understand that one. I was more confused by the custom reimplementation of 'then'.

**Pull request merged**
